File: hard/3753-total-waviness-of-numbers-in-range-ii/solution.java
/**
 * 3753. Total Waviness of Numbers in Range II
 * Time: O(D * 10 * 11 * 11 * 10)
 * Space: O(D * 11 * 11)
 */
class Solution {
    private char[] digits;
    private boolean[][][][] seen;
    private long[][][][] memoCount;
    private long[][][][] memoWaviness;

    public long totalWaviness(long num1, long num2) {
        return countUpTo(num2) - countUpTo(num1 - 1);
    }

    private long countUpTo(long limit) {
        if (limit <= 0) {
            return 0;
        }

        digits = Long.toString(limit).toCharArray();
        int n = digits.length;
        seen = new boolean[n][2][11][11];
        memoCount = new long[n][2][11][11];
        memoWaviness = new long[n][2][11][11];

        return dp(0, true, false, -1, -1).waviness;
    }

    private Result dp(
        int position,
        boolean tight,
        boolean started,
        int previousPreviousDigit,
        int previousDigit
    ) {
        if (position == digits.length) {
            return new Result(1, 0);
        }

        int startedIndex = started ? 1 : 0;
        int previousPreviousIndex = previousPreviousDigit + 1;
        int previousIndex = previousDigit + 1;

        if (!tight && seen[position][startedIndex][previousPreviousIndex][previousIndex]) {
            return new Result(
                memoCount[position][startedIndex][previousPreviousIndex][previousIndex],
                memoWaviness[position][startedIndex][previousPreviousIndex][previousIndex]
            );
        }

        int upperBound = tight ? digits[position] - '0' : 9;
        long totalCount = 0;
        long totalWaviness = 0;

        for (int digit = 0; digit <= upperBound; digit++) {
            boolean nextTight = tight && digit == upperBound;
            Result result;
            int contribution = 0;

            if (!started && digit == 0) {
                result = dp(position + 1, nextTight, false, -1, -1);
            } else if (!started) {
                result = dp(position + 1, nextTight, true, -1, digit);
            } else if (previousPreviousDigit == -1) {
                result = dp(position + 1, nextTight, true, previousDigit, digit);
            } else {
                if (
                    (previousPreviousDigit < previousDigit && previousDigit > digit)
                        || (previousPreviousDigit > previousDigit && previousDigit < digit)
                ) {
                    contribution = 1;
                }
                result = dp(position + 1, nextTight, true, previousDigit, digit);
            }

            totalCount += result.count;
            totalWaviness += result.waviness + (long) contribution * result.count;
        }

        if (!tight) {
            seen[position][startedIndex][previousPreviousIndex][previousIndex] = true;
            memoCount[position][startedIndex][previousPreviousIndex][previousIndex] = totalCount;
            memoWaviness[position][startedIndex][previousPreviousIndex][previousIndex] = totalWaviness;
        }

        return new Result(totalCount, totalWaviness);
    }

    private static class Result {
        long count;
        long waviness;

        Result(long count, long waviness) {
            this.count = count;
            this.waviness = waviness;
        }
    }
}

File: hard/3753-total-waviness-of-numbers-in-range-ii/solution.py
from functools import lru_cache
from typing import Tuple


class Solution:
    def totalWaviness(self, num1: int, num2: int) -> int:
        """
        3753. Total Waviness of Numbers in Range II
        Time: O(D * 10 * 11 * 11 * 10)
        Space: O(D * 11 * 11)
        """
        return self._count_up_to(num2) - self._count_up_to(num1 - 1)

    def _count_up_to(self, limit: int) -> int:
        if limit <= 0:
            return 0

        digits = list(map(int, str(limit)))

        @lru_cache(None)
        def dp(
            position: int,
            tight: bool,
            started: bool,
            previous_previous_digit: int,
            previous_digit: int,
        ) -> Tuple[int, int]:
            if position == len(digits):
                return 1, 0

            upper_bound = digits[position] if tight else 9
            total_numbers = 0
            total_waviness = 0

            for digit in range(upper_bound + 1):
                next_tight = tight and digit == upper_bound

                if not started and digit == 0:
                    count, waviness = dp(position + 1, next_tight, False, -1, -1)
                    total_numbers += count
                    total_waviness += waviness
                    continue

                if not started:
                    count, waviness = dp(position + 1, next_tight, True, -1, digit)
                    total_numbers += count
                    total_waviness += waviness
                    continue

                if previous_previous_digit == -1:
                    count, waviness = dp(
                        position + 1,
                        next_tight,
                        True,
                        previous_digit,
                        digit,
                    )
                    total_numbers += count
                    total_waviness += waviness
                    continue

                contribution = int(
                    previous_previous_digit < previous_digit > digit
                    or previous_previous_digit > previous_digit < digit
                )
                count, waviness = dp(position + 1, next_tight, True, previous_digit, digit)
                total_numbers += count
                total_waviness += waviness + contribution * count

            return total_numbers, total_waviness

        return dp(0, True, False, -1, -1)[1]
